### PR TITLE
std.experimental.allocator: makeArray should be safe for GCAllocator

### PR DESCRIPTION
**Pull request by wilzbach**
Hey after a lot of trail & error (#4225 #4213 #4190) I think I found a nice solution, how we can allow `makeArray` to be `@safe nothrow`.

My solution: assume `fillWithMemcpy` is `@safe` - everything else is then deduced by the D compiler.
I had to replace a couple of pointer occurrences with their `ref` version though.

As you can see if the range is not a ForwardRange it's not `@safe`, because `reallocate` isn't.
There is only one thing that I didn't like - casting is unsafe and hence I had to put the destroy method outside of the function and mark it as trusted.

What is your opinion @Hackerpilot @DmitryOlshansky @ZombineDev @JackStouffer ?

Btw can we use the same/similar approach for `make`?

### CODE REVIEW

File: std/experimental/allocator/gc_allocator.d [hunk 1]
```diff
@@ -80,7 +80,7 @@ struct GCAllocator
     }
 
     /// Ditto
-    @system bool deallocate(void[] b) shared
+    bool deallocate(void[] b) shared @trusted nothrow
```
Comment (JackStouffer) [inline std/experimental/allocator/gc_allocator.d]: Manually deallocating memory cannot and can never be `@trusted`.

Comment (wilzbach) [inline std/experimental/allocator/gc_allocator.d]: Okay then let's just check whether `makeArray` is `safe` ;-)
(updated)

File: std/experimental/allocator/package.d [hunk 2]
```diff
@@ -747,6 +749,11 @@ unittest
     test(theAllocator);
 }
 
+private void destroySafe(T)(auto ref T el) @trusted nothrow
```
Comment (DmitryOlshansky) [inline std/experimental/allocator/package.d]: Just use trusted  lambda in place, no point in extra module-level helpers

**Comment by DmitryOlshansky** [comment]
Makes sense to me

File: std/experimental/allocator/package.d [hunk 3]
```diff
@@ -760,19 +764,23 @@ if (isInputRange!R)
         auto result = cast(T[]) m;
 
         size_t i = 0;
-        scope (failure)
+
+        //scope (failure) alloc.deallocate(m);
```
Comment (wilzbach) [inline std/experimental/allocator/package.d]: if we really can't set deallocate with a GC to `@trusted`, then allocating with a forward range can't be safe either :/
Any ideas?

**Comment by JackStouffer** [comment]
With the changes to fillWithMemcpy and makeArray, the following works:

```
import std.experimental.allocator;
import std.stdio;

struct SafeGCAllocator
{
    import core.memory : GC;

    static shared SafeGCAllocator instance;
    @trusted shared void[] allocate(size_t bytes)
    {
        if (!bytes) return null;
        auto p = GC.malloc(bytes);
        return p ? p[0 .. bytes] : null;
    }
    size_t goodAllocSize(size_t n) shared @safe
    {
        if (n == 0)
            return 0;
        if (n <= 16)
            return 16;
        import core.bitop: bsr;

        auto largestBit = bsr(n-1) + 1;
        if (largestBit <= 12) // 4096 or less
            return size_t(1) << largestBit;

        return ((n + 4095) / 4096) * 4096;
    }
    // Will be collected via the GC, so no need to do anything
    @safe shared void deallocate(void[] b) {}
}

void main() @safe
{
    auto alloc = SafeGCAllocator.instance;

    int[] a = alloc.makeArray!int(2);
    scope(exit) alloc.dispose(a);
}
```

**Comment by wilzbach** [comment]
> With the changes to fillWithMemcpy and makeArray, the following works:

@JackStouffer yep you just remove `GC.free(b.ptr)` in deallocate (which I had initially) - so maybe I can repeat by question: What's wrong with setting this line to `@trusted`? We basically just help the GC?

> Deallocates the memory referenced by p. If p is null, no action occurs. If p references memory not originally allocated by this garbage collector, if p points to the interior of a memory block, or if this method is called from a finalizer, no action will be taken. The block will not be finalized regardless of whether the FINALIZE attribute is set. If finalization is desired, use delete instead.

**Comment by JackStouffer** [comment]
> What's wrong with setting this line to @trusted? We basically just help the GC?

Because you're manually freeing the memory which may be referenced by other things, rather than calling `GC.collect`.

**Comment by wilzbach** [comment]
@JackStouffer thanks for the explanation. Seems like OS X doesn't like me:

```
object.Exception@src/rt/minfo.d(167): Aborting: Cycle detected between modules with shared ctors/dtors:
std.experimental.allocator* ->
std.experimental.allocator.building_blocks.free_list ->
std.experimental.allocator.building_blocks.region* ->
std.experimental.allocator.mallocator ->
std.experimental.allocator
```

I didn't touch any import (or much in allocator at all) - any ideas what might be causing this problem?

**Comment by JackStouffer** [comment]
@wilzbach if you don't have a OS X VM you can use, you get to play the fun game of "What Makes the Auto-Tester Pass" by randomly commenting things out.

**Comment by wilzbach** [comment]
#3891 from @JakobOvrum seems to do the same and for `make` too - closing in favor of it.

**Pull request closed without merging**